File: 52. N-Queens II.cpp
class Solution {
private:
    int count = 0;
    int length;
    int* position;
    inline bool conflict(int n){
        for(register int i = 0; i < n; ++i){
            if(position[i] == position[n] || n - i == abs(position[i] - position[n])) return 1;
        }
        return 0;
    }
    inline void dfs(int n){
        if(n == length){
            ++count;
            return;
        }
        position[n] = 0;
        for(register int i = 0; i < length; ++i){
            if(!conflict(n)) dfs(n + 1);
            ++position[n];
        }
    }
public:
    int totalNQueens(int n) {
        length = n;
        position = new int[n];
        dfs(0);
        return count;
    }
};

File: 52_N-Queens_II.java
class Solution {

    private int result;

    public int totalNQueens(int n) {

        /* Sanity check */
        if(n == 0)
            return 0;

        result = 0;
        int[] pos = new int[n];

        dfs(pos, 0);

        return result;
    }

    /* dfs */
    private void dfs(int[] pos, int index) {
        /* One solution found */
        if(index == pos.length) {
            result++;
            return;
        }

        /* Keep searching */
        for(int i = 0; i < pos.length; ++i) {
            pos[index] = i;

            if(isConflict(pos, index) == false)
                dfs(pos, index + 1);
        }

    }

    /* isConflict - Returns whether column i conflicts with previous columns. */
    private boolean isConflict(int[] pos, int index) {
        for(int i = 0; i < index; ++i) {
            if(pos[index] == pos[i] || (index - i) == Math.abs(pos[index] - pos[i]))
                return true;
        }
        return false;
    }

}

